pkg/controller/channel/channel.go

```diff
@@ -0,0 +1,155 @@
+package channel
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path/filepath"
+	"time"
+
+	api "github.com/rancher/k3os/pkg/apis/k3os.io"
+	apiv1 "github.com/rancher/k3os/pkg/apis/k3os.io/v1"
+	"github.com/rancher/k3os/pkg/controller/upgradeset"
+	ctl "github.com/rancher/k3os/pkg/generated/controllers/k3os.io"
+	ctlv1 "github.com/rancher/k3os/pkg/generated/controllers/k3os.io/v1"
+	"github.com/rancher/norman/pkg/openapi"
+	"github.com/rancher/wrangler/pkg/apply"
+	"github.com/rancher/wrangler/pkg/crd"
+	"github.com/rancher/wrangler/pkg/generic"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	// DefaultPollingInterval is the default duration between attempts to resolve latest.
+	DefaultPollingInterval = 5 * time.Minute // TODO review default polling interval for channels
+)
+
+// BatchCreateCRD registers the Channel CRD
+func BatchCreateCRD(ctx context.Context, factory *crd.Factory) error {
+	prototype := apiv1.NewChannel("", "", apiv1.Channel{})
+	schema, err := openapi.ToOpenAPIFromStruct(*prototype)
+	if err != nil {
+		return err
+	}
+	factory.BatchCreateCRDs(ctx, crd.CRD{
+		GVK:        prototype.GroupVersionKind(),
+		PluralName: apiv1.ChannelResourceName,
+		Status:     true,
+		Schema:     schema,
+		Categories: []string{"all", "k3os", "upgrade"},
+		ShortNames: []string{"ch", "chs", "chan", "chans"},
+		//Columns: []v1beta1.CustomResourceColumnDefinition{
+		//	{
+		//		Name:     "Latest",
+		//		Type:     "string",
+		//		JSONPath: ".status.latestVersion",
+		//		Priority: 10,
+		//	},
+		//	{
+		//		Name:     "URL",
+		//		Type:     "string",
+		//		JSONPath: ".spec.url",
+		//		Priority: 100,
+		//	},
+		//},
+	})
+	return nil
+}
+
+type Options struct {
+	ControllerName    string
+	ControllerApply   apply.Apply
+	ControllerFactory *ctl.Factory
+	PollingInterval   time.Duration
+}
+
+// RegisterHandlers registers Channel handlers
+func RegisterHandlers(ctx context.Context, options Options) {
```

Comment:
An `Options` struct typically has an implied behavior that all fields are optional.  If something is required is should be a method argument.